src/hnsw/external_index.c

```diff
@@ -46,37 +49,69 @@ static char *extract_node(char               *data,
     return tape;
 }
 
-void StoreExternalIndex(Relation        index,
+void StoreExternalIndexBlockMapGroup(
+                        Relation        index,
                         usearch_index_t external_index,
                         ForkNumber      forkNum,
                         char           *data,
+                        int            *progress,
                         size_t          external_index_size,
                         int             dimension,
-                        size_t          num_added_vectors)
+                        int             first_node_index,
+                        size_t          num_added_vectors) 
 {
+    // Create empty blockmap pages for the group
+    const int number_of_blockmaps_in_group = 1 << g_blockmap_groups;
+    elog(INFO, "davkhech number_of_blockmaps_in_group %d num_added_vectors %d", number_of_blockmaps_in_group, num_added_vectors);
+    for (int blockmap_id = 0; blockmap_id < number_of_blockmaps_in_group; ++blockmap_id) {
+        Buffer buf = ReadBufferExtended(index, forkNum, P_NEW, RBM_NORMAL, NULL);
+        LockBuffer(buf, BUFFER_LOCK_EXCLUSIVE);
+
+        GenericXLogState *state = GenericXLogStart(index);
+        Page page = GenericXLogRegisterBuffer(state, buf, GENERIC_XLOG_FULL_IMAGE);
+        PageInit(page, BufferGetPageSize(buf), sizeof(HnswIndexPageSpecialBlock));
+
+        HnswIndexPageSpecialBlock *special = (HnswIndexPageSpecialBlock *)PageGetSpecialPointer(page);
+        special->firstId = first_node_index + blockmap_id * HNSW_BLOCKMAP_BLOCKS_PER_PAGE;
+        special->nextblockno = InvalidBlockNumber;
+
+        HnswBlockmapPage *blockmap = palloc0(BLCKSZ);
+
+        memset(blockmap, 0, BLCKSZ);
+        blockmap->first_id = first_node_index + blockmap_id * HNSW_BLOCKMAP_BLOCKS_PER_PAGE;
+
+        if(PageAddItem(page, (Item)blockmap, sizeof(HnswBlockmapPage), InvalidOffsetNumber, false, false)
+           == InvalidOffsetNumber) {
+            // we always add a single Blockmap page per Index page which has a fixed size that
+            // always fits in postgres wal page. So this should never happen
+            // (Assumes 8k BLKSZ. we can make HnswBlockmapPage size configurable by BLCKSZ)
+            elog(ERROR, "could not add blockmap to page");
+        }
+
+        special->lastId = first_node_index + (blockmap_id + 1) * HNSW_BLOCKMAP_BLOCKS_PER_PAGE - 1;
+        special->nextblockno = BufferGetBlockNumber(buf) + 1;
```

Comment:
Is this valid for the last iteration of the for loop?

Reply:
as we are going to create data blocks afterwards, it is valid